packages/console/src/pages/UserDetails/index.tsx

```diff
@@ -1,32 +1,101 @@
 import { User } from '@logto/schemas';
-import React from 'react';
+import React, { useEffect, useState } from 'react';
+import { useController, useForm } from 'react-hook-form';
+import { toast } from 'react-hot-toast';
 import { useTranslation } from 'react-i18next';
 import { useParams } from 'react-router-dom';
 import useSWR from 'swr';
 
 import BackLink from '@/components/BackLink';
+import Button from '@/components/Button';
 import Card from '@/components/Card';
+import CodeEditor from '@/components/CodeEditor';
 import CopyToClipboard from '@/components/CopyToClipboard';
+import FormField from '@/components/FormField';
 import ImagePlaceholder from '@/components/ImagePlaceholder';
-import { RequestError } from '@/hooks/use-api';
+import TabNav, { TabNavLink } from '@/components/TabNav';
+import TextInput from '@/components/TextInput';
+import useApi, { RequestError } from '@/hooks/use-api';
+import { safeParseJson } from '@/utilities/json';
 
 import CreateSuccess from './components/CreateSuccess';
 import * as styles from './index.module.scss';
 
+type FormData = {
+  primaryEmail: string;
+  primaryPhone: string;
+  username: string;
+  name: string;
+  avatar: string;
+  roles: string;
+  customData: string;
+};
+
 const UserDetails = () => {
   const { id } = useParams();
   const { t } = useTranslation(undefined, { keyPrefix: 'admin_console' });
 
-  const { data, error } = useSWR<User, RequestError>(id && `/api/users/${id}`);
+  const { data, error, mutate } = useSWR<User, RequestError>(id && `/api/users/${id}`);
   const isLoading = !data && !error;
 
+  const { handleSubmit, register, control, reset } = useForm<FormData>();
+  const [submitting, setSubmitting] = useState(false);
+  const {
+    field: { onChange, value },
+  } = useController({ name: 'customData', control, rules: { required: true } });
+  const api = useApi();
+
+  useEffect(() => {
+    if (!data) {
+      return;
+    }
+    reset({
+      primaryEmail: data.primaryEmail ?? '',
+      primaryPhone: data.primaryPhone ?? '',
+      username: data.username ?? '',
+      name: data.name ?? '',
+      avatar: data.avatar ?? '',
+      roles: data.roleNames.join(','),
+      customData: JSON.stringify(data.customData, null, 2),
```

Comment:
nit: prefer object destructuring?
```typescript
const {
  primaryEmail = '',
  primaryPhone = '',
  username = '',
  ...
} = data;
```

Reply:
Wait, can't we just directly set undefined to those fields?
```typescript
reset({
  ...data,
  roles: data.roleNames.join(','),
  customData: JSON.stringify(data.customData, null, 2),
})
```

Reply:
Per offline discussion, we will ignore this "null" problem for now, and discuss it in the next meeting. LOG-1928